specification/dartLangSpec.tex

```diff
@@ -22342,156 +24613,238 @@ \subsection{Operator Precedence}
+  if $T_0$ is a type variable $X_0$
+  or a promoted type variable \code{$X_0$\,\&\,$S_0$}
+  and $T_1$ is \code{$X_0$\,\&\,$S_1$}
+  then \SubtypeNE{T_0}{T_1} if{}f \SubtypeNE{T_0}{S_1}.
+
+  \commentary{%
+    Note that this rule is admissible, and can be safely elided if desired.%
+  }
+\item
+  \textbf{Right Promoted Variable:}
+  if $T_1$ is a promoted type variable \code{$X_1$\,\&\,$S_1$}
+  then \SubtypeNE{T_0}{T_1} if{}f \SubtypeNE{T_0}{X_1} and \SubtypeNE{T_0}{S_1}.
+\item
+  \textbf{Right FutureOr:}
+  if $T_1$ is \code{FutureOr<$S_1$>}
+  then \SubtypeNE{T_0}{T_1} if{}f any of the following hold:
+  \begin{itemize}
+  \item either \SubtypeNE{T_0}{\code{Future<$S_1$>}}.
+  \item or \SubtypeNE{T_0}{S_1}.
+  \item or $T_0$ is $X_0$ and $X_0$ has bound $B_0$ and \SubtypeNE{B_0}{T_1}.
+  \item or $T_0$ is \code{$X_0$\,\&\,$S_0$} and \SubtypeNE{S_0}{T_1}.
+  \end{itemize}
+\item
+  \textbf{Right Nullable:}
+  if $T_1$ is \code{$S_1$?}
+  then \SubtypeNE{T_0}{T_1} if{}f any of the following hold:
+  \begin{itemize}
+  \item either \SubtypeNE{T_0}{S_1}.
+  \item or \SubtypeNE{T_0}{\code{Null}}.
+  \item or $T_0$ is $X_0$ and $X_0$ has bound $B_0$ and \SubtypeNE{B_0}{T_1}.
+  \item or $T_0$ is \code{$X_0$\,\&\,$S_0$} and \SubtypeNE{S_0}{T_1}.
+  \end{itemize}
+\item
+  \textbf{Left Promoted Variable:}
+  $T_0$ is a promoted type variable \code{$X_0$\,\&\,$S_0$}
+  and \SubtypeNE{S_0}{T_1}.
+\item
+  \textbf{Left Type Variable Bound:}
+  $T_0$ is a type variable $X_0$ with bound $B_0$
+  and \SubtypeNE{B_0}{T_1}.
+\item
+  \textbf{Function Type/Function:}
+  $T_0$ is a function type and $T_1$ is \FUNCTION.
+\item
+  \textbf{Interface Compositionality:}
+  $T_0$ is an interface type \code{$C_0$<$S_0$, \ldots, $S_k$>}
+  and $T_1$ is \code{$C_0$<$U_0$, \ldots, $U_k$>} and each \SubtypeNE{S_i}{U_i}.
+\item
+  \textbf{Super-Interface:}
+  $T_0$ is an interface type with super-interfaces \List{S}{0}{n}
+  and \SubtypeNE{S_i}{T_1} for some $i$.
+\item
+  \textbf{Positional Function Types:}
+  $T_0$ is
+
+  \code{$U_0$ \FUNCTION<%
+    $X_0$\,\EXTENDS\,$B_{00}$, \ldots, $X_k$\,\EXTENDS\,$B_{0k}$>(%
+    $V_0$\,$x_0$, \ldots, $V_n$ $x_n$, %
+    [$V_{n+1}$\,\,$x_{n+1}$, \ldots, $V_m$\,\,$x_m$])}
+
+  and $T_1$ is
+
+  \code{$U_1$ \FUNCTION<%
+    $Y_0$\,\EXTENDS\,$B_{10}$, \ldots, $Y_k$\,\EXTENDS\,$B_{1k}$>(%
+    $S_0$\,$y_0$, \ldots, $S_p$\,$y_p$, %
+    [$S_{p+1}$\,$y_{p+1}$, \ldots, $S_q$\,$y_q$])}
+
+  such that each of the following criteria is satisfied,
+  where the $Z_i$ are fresh type variables with bounds
+  $B_{0i}[Z_0/X_0, \ldots, Z_k/X_k]$:
+
+  \begin{itemize}
+  \item $p \geq n$.
+  \item $m \geq q$.
+  \item \SubtypeNE{S_i[Z_0/Y_0, \ldots, Z_k/Y_k]}{V_i[Z_0/X_0, \ldots, Z_k/X_k]}
+    for $i \in 0 .. q$.
+  \item \SubtypeNE{U_0[Z_0/X_0, \ldots, Z_k/X_k]}{U_1[Z_0/Y_0, \ldots, Z_k/Y_k]}.
+  \item $B_{0i}[Z_0/X_0, \ldots, Z_k/X_k]$ and $B_{1i}[Z_0/Y_0, \ldots, Z_k/Y_k]$
+    have the same canonical syntax, for $i \in 0 .. k$.
```

Comment:
Should we ask for a weaker requirement: the bounds should be mutual subtypes or should be equal if all top types treated as being equal to each other and the bottom types treated as being equal to each other?

Reply:
Yes! The requirement should be 'mutual subtypes', and it's been updated now. I overlooked the change were `===` in the null safety feature spec was changed to mean 'mutual subtypes', and that's the reason why I kept the old text because `===` was still there. That shows the power of change propagation! ;-)